apps/src/blockly/addons/cdoSerializationHelpers.js

```diff
@@ -0,0 +1,184 @@
+import BlockSvgUnused from './blockSvgUnused';
+import {WORKSPACE_PADDING} from '../constants';
+import {frameSizes} from './cdoConstants';
+
+const {HEADER_HEIGHT, MARGIN_BOTTOM, MARGIN_SIDE, MARGIN_TOP} = frameSizes;
+const SVG_FRAME_HEIGHT = HEADER_HEIGHT + MARGIN_TOP + MARGIN_BOTTOM;
+const SVG_FRAME_SIDE_PADDING = MARGIN_SIDE;
+const SVG_FRAME_TOP_PADDING = HEADER_HEIGHT + MARGIN_TOP;
+const SORT_BY_POSITION = true;
+const VERTICAL_SPACE_BETWEEN_BLOCKS = 10;
+
+/**
+ * Converts XML serialization to JSON using a temporary unrendered workspace.
+ * @param {xml} xml - workspace serialization, current/legacy format
+ * @return {json} stateToLoad - modern workspace serialization
+ */
+export function convertXmlToJson(xml) {
+  const tempWorkspace = new Blockly.Workspace();
+
+  // domToBlockSpace returns an array of "block" objects with the following properties:
+  //   blockly_block: the actual block object created by Blockly
+  //   x: the x attribute found in <block/> element
+  //   y: the y attribute found in <block/> element
+  const xmlBlocks = Blockly.Xml.domToBlockSpace(tempWorkspace, xml);
+
+  const stateToLoad = Blockly.serialization.workspaces.save(tempWorkspace);
+
+  // Create a map of ids (key) and blocks (value).
+  const blockIdMap = stateToLoad.blocks.blocks.reduce(
+    (map, block) => map.set(block.id, block),
+    new Map()
+  );
+
+  addPositionsToState(xmlBlocks, blockIdMap);
+  tempWorkspace.dispose();
+  return stateToLoad;
+}
+
+/**
+ * Adds x/y values from XML to JSON serialization.
+ * @param {Array<Object>} xmlBlocks - an array of "block" objects containing a block and x/y coordinates
+ * @param {Map<String, Object>} blockIdMap - a map of ids (keys) and blocks (values)
+ */
+function addPositionsToState(xmlBlocks, blockIdMap) {
+  xmlBlocks.forEach(xmlBlock => {
+    const block = blockIdMap.get(xmlBlock.blockly_block.id);
+    if (block) {
+      // Do not change block values if xmlBlock values are NaN (unspecified in XML)
+      block.x = xmlBlock.x || block.x;
+      block.y = xmlBlock.y || block.y;
+    }
+  });
+}
+
+/**
+ * Position blocks on a workspace (if they do not already have positions)
+ * @param {Blockly.Workspace} workspace - the current Blockly workspace
+ * @param {function} [positionBlock] - moves a single block using the current cursor coordinates
+ * @param {Array<block>} [blocks] - an array of block objects
+ */
+export function positionBlocksOnWorkspace(
```

Comment:
I wonder if a comment may be helpful here for readers to understand that if `positionBlock` and `blocks` are passed, they are using the xml pathway as opposed to the json pathway which uses the default parameter values.